Run ecj in-process

dweiss opened the pull request:
This fixes #14911. I hope I got it right. Seems to be working and indeed, seems to be more snappy.

dweiss committed:
Run ecj in-process.

rmuir commented:
will try this one out real quick! awesome

dweiss commented:
This is what I get when I run this a few times:
```
> time ./gradlew cleanEcjLintMain ecjLintMain -Ptask.times=true
```

* on main:
```
Aggregate task times (possibly running in parallel!):
  45.75 sec.  ecjLintMain
   0.05 sec.  cleanEcjLintMain
   0.00 sec.  errorProneSkipped

BUILD SUCCESSFUL in 6s
153 actionable tasks: 75 executed, 78 up-to-date

real    0m6.473s
user    0m1.806s
sys     0m0.232s
```

* on this PR:
```
Aggregate task times (possibly running in parallel!):
   8.68 sec.  ecjLintMain
   0.11 sec.  cleanEcjLintMain
   0.00 sec.  errorProneSkipped

BUILD SUCCESSFUL in 4s
152 actionable tasks: 73 executed, 79 up-to-date

real    0m4.162s
user    0m1.695s
sys     0m0.239s
```

So the wall time isn't that much shorter (I do have a multi-core machine though). If I add ```--max-workers 2```, the results show the speedup:
* main:
```
BUILD SUCCESSFUL in 22s
153 actionable tasks: 75 executed, 78 up-to-date

real    0m23.037s
user    0m1.717s
sys	0m0.209s
```
* PR:
```
BUILD SUCCESSFUL in 5s
152 actionable tasks: 73 executed, 79 up-to-date

real    0m6.015s
user    0m1.748s
sys	0m0.261s
```

dweiss commented:
Take your time. I also updated ECJ to the maven central version. I'm going to sleep now -- feel free to play around. I think the time savings do justify running ecj in-process with gradle. If it explodes for some reason (daemon heap), we can think of alternatives like running it using a gradle service (which can be a separate JVM but where parallelism can be kept within reasonable limits).

rmuir commented:
I can also confirm the linter is still actually linting and functioning correctly (when they get so much faster... you know). I inserted a bug that only ECJ will find:

```
> Task :lucene:core:ecjLintTest FAILED

[Incubating] Problems report is available at: file:///home/rmuir/workspace/lucene/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':lucene:core:ecjLintTest'.
> ecj linter failed:
  ----------
  1. ERROR in /home/rmuir/workspace/lucene/lucene/core/src/test/org/apache/lucene/TestDemo.java (at line 52)
        * @throws
           ^^^^^^
  Javadoc: Missing class name
  ----------
  1 problem (1 error)


* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 32s
707 actionable tasks: 179 executed, 528 up-to-date
```

rmuir committed:
style: add a type-hint to deconfuse eclipse/jdtls

eclipse/language server doesn't seem to perfectly do the type inference
with var, sometimes it yields false errors because of it.

Looking at their git repo, they seem to be heavily reworking this (just
yesterday). But adding this one type fixes the false-errors for now.

rmuir reviewed (approved):
I'm not really qualified to review it, but this is working great from my testing. I really like that logic is moved to java as well.

Thank you!
